staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go

```diff
@@ -117,12 +190,223 @@ func (a *ApplySet) addLabels(objects []*resource.Info) error {
+		if err != nil {
+			return nil, fmt.Errorf("could not find kind for resource in %q annotation: %w", ApplySetGRsAnnotation, err)
+		}
+		mappings[mapping.Resource] = mapping
+	}
+	return mappings, nil
+}
+
+func parseNamespacesAnnotation(annotations map[string]string) sets.Set[string] {
+	namespaces := make(sets.Set[string])
+	annotation, ok := annotations[ApplySetAdditionalNamespacesAnnotation]
+	if !ok { // this annotation is completely optional
+		return namespaces
+	}
+	for _, ns := range strings.Split(annotation, ",") {
+		namespaces.Insert(ns)
+	}
+	return namespaces
+}
+
+type ApplySetUpdateMode string
+
+var SetUpdateSuccessful ApplySetUpdateMode = "latest"
+var SetUpdateIncomplete ApplySetUpdateMode = "superset"
+
+func (a *ApplySet) UpdateParent(mode ApplySetUpdateMode, dryRun cmdutil.DryRunStrategy, validation string) error {
+	data, err := json.Marshal(a.buildParentPatch(mode))
+	if err != nil {
+		return fmt.Errorf("failed to encode patch for ApplySet parent: %w", err)
+	}
+	err = serverSideApplyRequest(a, data, dryRun, validation, false)
+	if err != nil && errors.IsConflict(err) {
+		// Try again with conflicts forced
+		klog.Warningf("WARNING: failed to update ApplySet: %s\nApplySet field manager %s should own these fields. Retrying with conflicts forced.", err.Error(), a.FieldManager())
+		err = serverSideApplyRequest(a, data, dryRun, validation, true)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to update ApplySet: %w", err)
+	}
+	return nil
+}
+
+func serverSideApplyRequest(a *ApplySet, data []byte, dryRun cmdutil.DryRunStrategy, validation string, forceConficts bool) error {
+	if dryRun == cmdutil.DryRunClient {
+		return nil
+	}
+	helper := resource.NewHelper(a.client, a.parent.RESTMapping).
+		DryRun(dryRun == cmdutil.DryRunServer).
+		WithFieldManager(a.FieldManager()).
+		WithFieldValidation(validation)
+
+	options := metav1.PatchOptions{
+		Force: &forceConficts,
+	}
+	_, err := helper.Patch(
+		a.parent.Namespace,
+		a.parent.Name,
+		types.ApplyPatchType,
+		data,
+		&options,
+	)
+	return err
+}
+
+func (a *ApplySet) buildParentPatch(mode ApplySetUpdateMode) *metav1.PartialObjectMetadata {
+	var newGRsAnnotation, newNsAnnotation string
+	switch mode {
+	case SetUpdateIncomplete:
+		// If the apply succeeded but pruning failed, the set of group resources that
+		// the ApplySet should track is the superset of the previous and current resources.
+		// This ensures that the resources that failed to be pruned are not orphaned from the set.
+		grSuperset := make(map[schema.GroupVersionResource]*meta.RESTMapping)
+		for versionResource, mapping := range a.currentResources {
+			grSuperset[versionResource] = mapping
+		}
+		for versionResource, mapping := range a.updatedResources {
+			grSuperset[versionResource] = mapping
+		}
+		newGRsAnnotation = generateResourcesAnnotation(grSuperset)
+		newNsAnnotation = generateNamespacesAnnotation(a.currentNamespaces.Union(a.updatedNamespaces), a.parent.Namespace)
```

Comment:
Cool that we can use union here!

Makes me wonder if there's a function to go from `map[K]V` to `Set[K]`, but ... another PR!

Edit: there is - `sets.KeySet` - cool... TIL!